### PR TITLE
add uri template to the package

### PR DESCRIPTION
**Pull request by gmponos**

### CODE REVIEW

**Comment by gmponos** [comment]
Hi.. are you interested into moving this in here?

**Comment by Nyholm** [comment]
What is the use case?

**Comment by GrahamCampbell** [comment]
Using it outside of Guzzle I guess?

File: src/UriTemplate.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+declare(strict_types=1);
+
+namespace GuzzleHttp\Psr7;
+
+/**
+ * Expands URI templates. Userland implementation of PECL uri_template.
+ *
+ * @link http://tools.ietf.org/html/rfc6570
+ */
+class UriTemplate
```
Comment (GrahamCampbell) [inline src/UriTemplate.php:12]: final?

File: src/UriTemplate.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            $this->template
+        );
+    }
+
+    /**
+     * Parse an expression into parts
+     *
+     * @param string $expression Expression to parse
+     *
+     * @return array Returns an associative array of parts
```
Comment (GrahamCampbell) [inline src/UriTemplate.php:100]: can we be more specific than array?

File: src/UriTemplate.php [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            $result['values'][] = $varspec;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Process an expansion
+     *
+     * @param array $matches Matches met in the preg_replace_callback
```
Comment (GrahamCampbell) [inline src/UriTemplate.php:136]: `string[]`?

File: src/UriTemplate.php [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+     * Determines if an array is associative.
+     *
+     * This makes the assumption that input arrays are sequences or hashes.
+     * This assumption is a tradeoff for accuracy in favor of speed, but it
+     * should work in almost every case where input is supplied for a URI
+     * template.
+     *
+     * @param array $array Array to check
+     */
+    private function isAssoc(array $array): bool
```
Comment (GrahamCampbell) [inline src/UriTemplate.php:259]: why return typing here but not in the other methods?

**Comment by gmponos** [comment]
> What is the use case?

I have various projects/packages that I am using it. So once guzzle v7 is out.. So my migration path is a little bit painful.

**Comment by gmponos** [comment]
btw this PR is a followup of this https://github.com/guzzle/guzzle/pull/2440

**Comment by gmponos** [comment]
> I have various projects/packages that I am using it. So once guzzle v7 is out.. my migration path will be a little bit painful.

so I just saw on the PR linked above that the phpleague added a UriTemplate.. you could say that someone could use that one.. but it feels weird though that we had to drop ours and I need to install an external package to migrate..

File: src/UriTemplate.php [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        '%2B',
+        '%2C',
+        '%3B',
+        '%3D'
+    ];
+
+    /**
+     * @return string|string[]|null
+     */
+    public function expand(string $template, array $variables)
```
[the remark this answers is not in the document]
Comment (gmponos) [reply src/UriTemplate.php]: done.. feedback is welcome.. :)

[the remark this answers is not in the document]
Comment (ro0NL) [reply src/UriTemplate.php]: LGTM :sweat_smile:

**Comment by ro0NL** [comment]
ive no strong opionion on a separate package, but doesnt anything evolving around URIs more or less fit here?  There's UriResolver, UriNormalizer.. and now UriTemplate :D

**Comment by gmponos** [comment]
> There's UriResolver, UriNormalizer.. and now UriTemplate

That's what drove me submitting a PR here.. 

Neither I  have strong arguments here...

I wouldn't mind creating a package separate package under guzzle org related with it.. @Nyholm @sagikazarmark WDYT?

**Comment by gmponos** [comment]
Created this repo.. https://github.com/guzzle/uri-template

will try to during the weekend to move the code there.. 

Until then if you have any objections/concerns let me know.

**Comment by GrahamCampbell** [comment]
I assume the min php version for that repo will be 7.2?

**Comment by gmponos** [comment]
Yes.. that's my intention

**Comment by GrahamCampbell** [comment]
Sounds good to me. 👍

**Comment by GrahamCampbell** [comment]
This PR can be closed now?

**Pull request closed without merging**